File: go/src/backtest_engine.go
// Backtest Engine — Go
// Tests a rule-based trading strategy against historical OHLCV bars (no money
// at risk — pure computation). Mirrors backtest_engine.py exactly in behaviour.

package main

import (
	"fmt"
	"math"
)

type Signal string

const (
	SignalLong  Signal = "long"
	SignalShort Signal = "short"
	SignalFlat  Signal = "flat"
)

type SignalFn func(history []OHLCVBar) Signal

type BacktestError struct{ Msg string }

func (e *BacktestError) Error() string { return e.Msg }

func btErrorf(format string, args ...interface{}) error {
	return &BacktestError{Msg: fmt.Sprintf(format, args...)}
}

type BTTrade struct {
	EntryDate  string  `json:"entry_date"`
	ExitDate   string  `json:"exit_date"`
	Direction  Signal  `json:"direction"`
	EntryPrice float64 `json:"entry_price"`
	ExitPrice  float64 `json:"exit_price"`
	PnlPct     float64 `json:"pnl_pct"`
	BarsHeld   int     `json:"bars_held"`
}

type BacktestResult struct {
	TotalReturnPct      float64   `json:"total_return_pct"`
	AnnualizedReturnPct float64   `json:"annualized_return_pct"`
	SharpeRatio         float64   `json:"sharpe_ratio"`
	MaxDrawdownPct      float64   `json:"max_drawdown_pct"`
	WinRatePct          float64   `json:"win_rate_pct"`
	NumTrades           int       `json:"num_trades"`
	AvgTradePct         float64   `json:"avg_trade_pct"`
	BestTradePct        float64   `json:"best_trade_pct"`
	WorstTradePct       float64   `json:"worst_trade_pct"`
	Trades              []BTTrade `json:"trades"`
	EquityCurve         []float64 `json:"equity_curve"`
}

type BacktestOptions struct {
	InitialCapital     float64
	FeePct             float64
	RiskFreeRateAnnual float64
	BarsPerYear        int
}

func DefaultBacktestOptions() BacktestOptions {
	return BacktestOptions{InitialCapital: 10_000, FeePct: 0.001, RiskFreeRateAnnual: 0, BarsPerYear: 252}
}

func RunBacktest(bars []OHLCVBar, signalFn SignalFn, opts BacktestOptions) (*BacktestResult, error) {
	if opts.InitialCapital == 0 {
		opts.InitialCapital = 10_000
	}
	if opts.BarsPerYear == 0 {
		opts.BarsPerYear = 252
	}

	if len(bars) < 2 {
		return nil, btErrorf("need at least 2 bars to run a backtest")
	}

	equity := opts.InitialCapital
	equityCurve := []float64{equity}
	position := SignalFlat
	entryPrice := 0.0
	entryIdx := 0
	var trades []BTTrade
	var returns []float64

	for i := 1; i < len(bars); i++ {
		history := bars[:i+1]
		newSignal := signalFn(history)
		if newSignal != SignalLong && newSignal != SignalShort && newSignal != SignalFlat {
			return nil, btErrorf("signalFn returned invalid signal '%s' at bar %d", newSignal, i)
		}

		currentClose := bars[i].Close
		prevClose := bars[i-1].Close

		barReturn := 0.0
		switch position {
		case SignalLong:
			barReturn = (currentClose - prevClose) / prevClose
		case SignalShort:
			barReturn = (prevClose - currentClose) / prevClose
		}
		equity *= 1 + barReturn
		returns = append(returns, barReturn)
		equityCurve = append(equityCurve, equity)

		if newSignal != position {
			if position != SignalFlat {
				exitPrice := currentClose
				var pnlPct float64
				if position == SignalLong {
					pnlPct = (exitPrice - entryPrice) / entryPrice * 100
				} else {
					pnlPct = (entryPrice - exitPrice) / entryPrice * 100
				}
				pnlPct -= opts.FeePct * 100
				equity *= 1 - opts.FeePct
				trades = append(trades, BTTrade{
					EntryDate: bars[entryIdx].Date, ExitDate: bars[i].Date,
					Direction: position, EntryPrice: entryPrice, ExitPrice: exitPrice,
					PnlPct: pnlPct, BarsHeld: i - entryIdx,
				})
			}
			if newSignal != SignalFlat {
				entryPrice = currentClose
				entryIdx = i
				equity *= 1 - opts.FeePct
			}
			position = newSignal
		}
	}

	if position != SignalFlat {
		exitPrice := bars[len(bars)-1].Close
		var pnlPct float64
		if position == SignalLong {
			pnlPct = (exitPrice - entryPrice) / entryPrice * 100
		} else {
			pnlPct = (entryPrice - exitPrice) / entryPrice * 100
		}
		pnlPct -= opts.FeePct * 100
		trades = append(trades, BTTrade{
			EntryDate: bars[entryIdx].Date, ExitDate: bars[len(bars)-1].Date,
			Direction: position, EntryPrice: entryPrice, ExitPrice: exitPrice,
			PnlPct: pnlPct, BarsHeld: len(bars) - 1 - entryIdx,
		})
	}

	totalReturnPct := (equityCurve[len(equityCurve)-1] - opts.InitialCapital) / opts.InitialCapital * 100
	numPeriods := len(bars) - 1
	years := float64(numPeriods) / float64(opts.BarsPerYear)
	finalEquity := equityCurve[len(equityCurve)-1]
	annualizedReturnPct := 0.0
	if years > 0 && finalEquity > 0 {
		annualizedReturnPct = (math.Pow(finalEquity/opts.InitialCapital, 1/years) - 1) * 100
	}

	sharpe := computeSharpe(returns, opts.RiskFreeRateAnnual, opts.BarsPerYear)
	maxDD := computeMaxDrawdown(equityCurve)

	var winCount int
	var sumPnl, best, worst float64
	if len(trades) > 0 {
		best = trades[0].PnlPct
		worst = trades[0].PnlPct
	}
	for _, t := range trades {
		if t.PnlPct > 0 {
			winCount++
		}
		sumPnl += t.PnlPct
		if t.PnlPct > best {
			best = t.PnlPct
		}
		if t.PnlPct < worst {
			worst = t.PnlPct
		}
	}
	winRate := 0.0
	avgTrade := 0.0
	if len(trades) > 0 {
		winRate = float64(winCount) / float64(len(trades)) * 100
		avgTrade = sumPnl / float64(len(trades))
	}

	return &BacktestResult{
		TotalReturnPct: totalReturnPct, AnnualizedReturnPct: annualizedReturnPct,
		SharpeRatio: sharpe, MaxDrawdownPct: maxDD, WinRatePct: winRate,
		NumTrades: len(trades), AvgTradePct: avgTrade, BestTradePct: best, WorstTradePct: worst,
		Trades: trades, EquityCurve: equityCurve,
	}, nil
}

func computeSharpe(returns []float64, riskFreeRateAnnual float64, barsPerYear int) float64 {
	if len(returns) < 2 {
		return 0
	}
	mean := 0.0
	for _, r := range returns {
		mean += r
	}
	mean /= float64(len(returns))

	variance := 0.0
	for _, r := range returns {
		variance += (r - mean) * (r - mean)
	}
	variance /= float64(len(returns) - 1)
	stdDev := math.Sqrt(variance)
	if stdDev == 0 {
		return 0
	}
	riskFreePerPeriod := riskFreeRateAnnual / float64(barsPerYear)
	return (mean - riskFreePerPeriod) / stdDev * math.Sqrt(float64(barsPerYear))
}

func computeMaxDrawdown(equityCurve []float64) float64 {
	if len(equityCurve) < 2 {
		return 0
	}
	peak := equityCurve[0]
	maxDD := 0.0
	for _, v := range equityCurve {
		if v > peak {
			peak = v
		}
		dd := 0.0
		if peak > 0 {
			dd = (peak - v) / peak
		}
		if dd > maxDD {
			maxDD = dd
		}
	}
	return maxDD * 100
}

// ── Built-in reference strategies ───────────────────────────────────────────────
func MakeSMACrossoverSignal(fastPeriod, slowPeriod int) SignalFn {
	if fastPeriod == 0 {
		fastPeriod = 50
	}
	if slowPeriod == 0 {
		slowPeriod = 200
	}
	return func(history []OHLCVBar) Signal {
		if len(history) < slowPeriod {
			return SignalFlat
		}
		closes := make([]float64, len(history))
		for i, b := range history {
			closes[i] = b.Close
		}
		fast := computeSMA(closes, fastPeriod)
		slow := computeSMA(closes, slowPeriod)
		f := fast[len(fast)-1]
		s := slow[len(slow)-1]
		if f == nil || s == nil {
			return SignalFlat
		}
		if *f > *s {
			return SignalLong
		}
		return SignalFlat
	}
}

func MakeRSIMeanReversionSignal(period int, oversold, overbought float64) SignalFn {
	if period == 0 {
		period = 14
	}
	if oversold == 0 {
		oversold = 30
	}
	if overbought == 0 {
		overbought = 70
	}
	return func(history []OHLCVBar) Signal {
		if len(history) < period+1 {
			return SignalFlat
		}
		closes := make([]float64, len(history))
		for i, b := range history {
			closes[i] = b.Close
		}
		rsi := computeRSI(closes, period)
		r := rsi[len(rsi)-1]
		if r == nil {
			return SignalFlat
		}
		if *r < oversold {
			return SignalLong
		}
		return SignalFlat
	}
}

func MakeBreakoutSignal(lookback int) SignalFn {
	if lookback == 0 {
		lookback = 20
	}
	return func(history []OHLCVBar) Signal {
		if len(history) < lookback+1 {
			return SignalFlat
		}
		window := history[len(history)-lookback-1 : len(history)-1]
		rangeHigh := window[0].High
		for _, b := range window {
			if b.High > rangeHigh {
				rangeHigh = b.High
			}
		}
		if history[len(history)-1].Close > rangeHigh {
			return SignalLong
		}
		return SignalFlat
	}
}

File: java/src/BacktestEngine.java
/**
 * Backtest Engine — Java
 * Tests a rule-based trading strategy against historical OHLCV bars.
 * No money at risk — pure computation. Mirrors backtest_engine.py exactly.
 */

import java.util.*;
import java.util.function.Function;

public class BacktestEngine {

    public enum Signal { LONG, SHORT, FLAT }

    @FunctionalInterface
    public interface SignalFn extends Function<List<MarketData.OHLCVBar>, Signal> {}

    public record Trade(
        String entryDate, String exitDate, Signal direction,
        double entryPrice, double exitPrice, double pnlPct, int barsHeld
    ) {}

    public record BacktestResult(
        double totalReturnPct,
        double annualizedReturnPct,
        double sharpeRatio,
        double maxDrawdownPct,
        double winRatePct,
        int numTrades,
        double avgTradePct,
        double bestTradePct,
        double worstTradePct,
        List<Trade> trades,
        List<Double> equityCurve
    ) {}

    public record BacktestOptions(
        double initialCapital,
        double feePct,
        double riskFreeRateAnnual,
        int barsPerYear
    ) {
        public static BacktestOptions defaults() {
            return new BacktestOptions(10_000.0, 0.001, 0.0, 252);
        }
        public static BacktestOptions withBarsPerYear(int bpy) {
            return new BacktestOptions(10_000.0, 0.001, 0.0, bpy);
        }
    }

    public static BacktestResult runBacktest(
        List<MarketData.OHLCVBar> bars,
        SignalFn signalFn,
        BacktestOptions opts
    ) {
        if (bars.size() < 2)
            throw new IllegalArgumentException("Need at least 2 bars to run a backtest");

        double equity = opts.initialCapital();
        List<Double> equityCurve = new ArrayList<>();
        equityCurve.add(equity);
        Signal position = Signal.FLAT;
        double entryPrice = 0.0;
        int entryIdx = 0;
        List<Trade> trades = new ArrayList<>();
        List<Double> returns = new ArrayList<>();

        for (int i = 1; i < bars.size(); i++) {
            List<MarketData.OHLCVBar> history = bars.subList(0, i + 1);
            Signal newSignal = signalFn.apply(history);
            if (newSignal == null) newSignal = Signal.FLAT;

            double currentClose = bars.get(i).close();
            double prevClose    = bars.get(i - 1).close();

            double barReturn = switch (position) {
                case LONG  -> (currentClose - prevClose) / prevClose;
                case SHORT -> (prevClose - currentClose) / prevClose;
                case FLAT  -> 0.0;
            };
            equity *= (1.0 + barReturn);
            returns.add(barReturn);
            equityCurve.add(equity);

            if (newSignal != position) {
                if (position != Signal.FLAT) {
                    double pnlPct = switch (position) {
                        case LONG  -> (currentClose - entryPrice) / entryPrice * 100.0;
                        case SHORT -> (entryPrice - currentClose) / entryPrice * 100.0;
                        case FLAT  -> 0.0;
                    };
                    pnlPct -= opts.feePct() * 100.0;
                    equity *= (1.0 - opts.feePct());
                    trades.add(new Trade(
                        bars.get(entryIdx).date(), bars.get(i).date(), position,
                        entryPrice, currentClose, pnlPct, i - entryIdx
                    ));
                }
                if (newSignal != Signal.FLAT) {
                    entryPrice = currentClose;
                    entryIdx   = i;
                    equity *= (1.0 - opts.feePct());
                }
                position = newSignal;
            }
        }

        // Close open position at final bar
        if (position != Signal.FLAT) {
            double exitPrice = bars.get(bars.size() - 1).close();
            double pnlPct = switch (position) {
                case LONG  -> (exitPrice - entryPrice) / entryPrice * 100.0;
                case SHORT -> (entryPrice - exitPrice) / entryPrice * 100.0;
                case FLAT  -> 0.0;
            };
            pnlPct -= opts.feePct() * 100.0;
            trades.add(new Trade(
                bars.get(entryIdx).date(), bars.get(bars.size() - 1).date(), position,
                entryPrice, exitPrice, pnlPct, bars.size() - 1 - entryIdx
            ));
        }

        double finalEquity     = equityCurve.get(equityCurve.size() - 1);
        double totalReturnPct  = (finalEquity - opts.initialCapital()) / opts.initialCapital() * 100.0;
        double years           = (bars.size() - 1.0) / opts.barsPerYear();
        double annualizedReturnPct = (years > 0 && finalEquity > 0)
            ? (Math.pow(finalEquity / opts.initialCapital(), 1.0 / years) - 1.0) * 100.0
            : 0.0;

        double sharpeRatio     = computeSharpe(returns, opts.riskFreeRateAnnual(), opts.barsPerYear());
        double maxDrawdownPct  = computeMaxDrawdown(equityCurve);

        long winCount = trades.stream().filter(t -> t.pnlPct() > 0).count();
        double winRatePct = trades.isEmpty() ? 0.0 : (double) winCount / trades.size() * 100.0;
        double avgTradePct = trades.isEmpty() ? 0.0 : trades.stream().mapToDouble(Trade::pnlPct).average().orElse(0.0);
        double bestTradePct  = trades.isEmpty() ? 0.0 : trades.stream().mapToDouble(Trade::pnlPct).max().orElse(0.0);
        double worstTradePct = trades.isEmpty() ? 0.0 : trades.stream().mapToDouble(Trade::pnlPct).min().orElse(0.0);

        return new BacktestResult(
            totalReturnPct, annualizedReturnPct, sharpeRatio, maxDrawdownPct,
            winRatePct, trades.size(), avgTradePct, bestTradePct, worstTradePct,
            trades, equityCurve
        );
    }

    public static double computeSharpe(List<Double> returns, double riskFreeRateAnnual, int barsPerYear) {
        if (returns.size() < 2) return 0.0;
        double mean = returns.stream().mapToDouble(Double::doubleValue).average().orElse(0.0);
        double variance = returns.stream().mapToDouble(r -> (r - mean) * (r - mean)).sum() / (returns.size() - 1);
        double stdDev = Math.sqrt(variance);
        if (stdDev == 0.0) return 0.0;
        double riskFreePerPeriod = riskFreeRateAnnual / barsPerYear;
        return (mean - riskFreePerPeriod) / stdDev * Math.sqrt(barsPerYear);
    }

    public static double computeMaxDrawdown(List<Double> equityCurve) {
        if (equityCurve.size() < 2) return 0.0;
        double peak = equityCurve.get(0), maxDd = 0.0;
        for (double v : equityCurve) {
            if (v > peak) peak = v;
            double dd = peak > 0 ? (peak - v) / peak : 0.0;
            if (dd > maxDd) maxDd = dd;
        }
        return maxDd * 100.0;
    }

    // ── Built-in reference strategies ─────────────────────────────────────────
    /** Golden/death-cross momentum. See knowledge/trading_strategies/momentum_strategies.txt */
    public static SignalFn makeSMACrossoverSignal(int fastPeriod, int slowPeriod) {
        int fast = fastPeriod <= 0 ? 50  : fastPeriod;
        int slow = slowPeriod <= 0 ? 200 : slowPeriod;
        return history -> {
            if (history.size() < slow) return Signal.FLAT;
            double[] closes = history.stream().mapToDouble(MarketData.OHLCVBar::close).toArray();
            double[] fastSMA = MarketData.computeSMA(closes, fast);
            double[] slowSMA = MarketData.computeSMA(closes, slow);
            int last = closes.length - 1;
            if (Double.isNaN(fastSMA[last]) || Double.isNaN(slowSMA[last])) return Signal.FLAT;
            return fastSMA[last] > slowSMA[last] ? Signal.LONG : Signal.FLAT;
        };
    }

    /** RSI oversold bounce. See knowledge/trading_strategies/mean_reversion_strategies.txt */
    public static SignalFn makeRSIMeanReversionSignal(int period, double oversold, double overbought) {
        int p  = period <= 0 ? 14 : period;
        double os = oversold <= 0 ? 30.0 : oversold;
        return history -> {
            if (history.size() < p + 1) return Signal.FLAT;
            double[] closes = history.stream().mapToDouble(MarketData.OHLCVBar::close).toArray();
            double[] rsi = MarketData.computeRSI(closes, p);
            double last = rsi[rsi.length - 1];
            if (Double.isNaN(last)) return Signal.FLAT;
            return last < os ? Signal.LONG : Signal.FLAT;
        };
    }

    /** Range breakout. See knowledge/trading_strategies/breakout_strategies.txt */
    public static SignalFn makeBreakoutSignal(int lookback) {
        int lb = lookback <= 0 ? 20 : lookback;
        return history -> {
            if (history.size() < lb + 1) return Signal.FLAT;
            List<MarketData.OHLCVBar> window = history.subList(history.size() - lb - 1, history.size() - 1);
            double rangeHigh = window.stream().mapToDouble(MarketData.OHLCVBar::high).max().orElse(0.0);
            return history.get(history.size() - 1).close() > rangeHigh ? Signal.LONG : Signal.FLAT;
        };
    }
}
